### ISSUE TITLE
Segmentation doesn't show up in SlicerDICOMDatabase after creation

### ISSUE THREAD

**Issue by che85** [QIICR/QuantitativeReporting#170]
And using advanced mode proposes DICOMSegmentation for MR image instead of ScalarVolume

![image](https://user-images.githubusercontent.com/10195822/28528774-a6ff7808-705c-11e7-81f7-c6beba5c2b82.png)

Trying to load MR as ScalarVolume gives the following error:
![image](https://user-images.githubusercontent.com/10195822/28528842-dfe8360a-705c-11e7-9842-b3f4b8caf56a.png)

Not sure what's going on, but one thing I noticed is that I can load the MR image before the segmentation was created. After creation and restarting Slicer the just mentioned situation comes up.

**Comment by fedorov** [comment]
Interesting. I wonder if this is related to the DCMTK upgrade somehow.

Related to #169!

**Comment by che85** [comment]
Yeah it's either the DCMTK update or because we are using the master of dcmqi now for the extension

**Comment by pieper** [comment]
That's interesting - I added exception handling to the vtkITK readers that GDCM wouldn't end up triggering the uncaught error dialog.  Let me know if there's a way to replicate this issue.

**Comment by che85** [comment]
@pieper 
1. download the nightly Slicer
2. install `Quantitative Reporting`
3. manually add the DICOM plugins of Slicer-PETDICOMExtension[1] to your additional module paths
4. restart
5. open `Quantitative Reporting` and click button `Retrieve and load test data`
6. create segmentation
7. click `Save Report` 

Then you should see the SR in your SlicerDICOMDatabase, but no Seg shows up and when trying to load the MR as a ScalarVolume you get the errors mentioned earlier.


[1] https://github.com/QIICR/Slicer-PETDICOMExtension

**Comment by cpinter** [comment]
Not sure if I understand the problem correctly, but the DICOM SEG export plugin is broken, see https://github.com/QIICR/QuantitativeReporting/issues/140

I was going to fix it this week. It's somewhat more complex than a simple SH 2.0 porting due to the fact that terminologies are now not stored in the color logic, as it is still like that in the current state of the plugin.

**Comment by fedorov** [comment]
@cpinter it looks like the import plugin is now broken too!

**Comment by cpinter** [comment]
Oh I see! OK that might be something new.

**Comment by fedorov** [comment]
I am also getting this error on startup - @pieper can this be related to your refactoring of the DICOM load functionality?

```Traceback (most recent call last):
  File "/Applications/Slicer.app/Contents/lib/Slicer-4.7/qt-scripted-modules/DICOM.py", line 85, in performPostModuleDiscoveryTasks
    DICOMLib.setDatabasePrecacheTags()
  File "/Applications/Slicer.app/Contents/lib/Slicer-4.7/qt-scripted-modules/DICOMLib/DICOMWidgets.py", line 36, in setDatabasePrecacheTags
    plugin = slicer.modules.dicomPlugins[pluginClass]()
  File "/Applications/Slicer.app/Contents/Extensions-26167/QuantitativeReporting/lib/Slicer-4.7/qt-scripted-modules/DICOMTID1500Plugin.py", line 23, in __init__
    self.rwvmPlugin = slicer.modules.dicomPlugins["DICOMRWVMPlugin"]()
KeyError: 'DICOMRWVMPlugin'
```

**Comment by fedorov** [comment]
related to the above, here is the full log - RWVM plugin failed to load, but no hints as to why.

```[DEBUG][Qt] 24.07.2017 11:11:00 [] (unknown:0) - Session start time .......: 2017-07-24 11:11:00
[DEBUG][Qt] 24.07.2017 11:11:00 [] (unknown:0) - Slicer version ...........: 4.7.0-2017-07-22 (revision 26167) macosx-amd64 - installed
[DEBUG][Qt] 24.07.2017 11:11:00 [] (unknown:0) - Operating system .........: Mac OS X / 10.12.5 / 16F73 - 64-bit
[DEBUG][Qt] 24.07.2017 11:11:00 [] (unknown:0) - Memory ...................: 16384 MB physical, 1024 MB virtual
[DEBUG][Qt] 24.07.2017 11:11:00 [] (unknown:0) - CPU ......................: GenuineIntel Intel(R) Core(TM) i7-4980HQ CPU @ 2.80GHz, 4 cores, 8 logical processors
[DEBUG][Qt] 24.07.2017 11:11:00 [] (unknown:0) - Developer mode enabled ...: yes
[DEBUG][Qt] 24.07.2017 11:11:00 [] (unknown:0) - Prefer executable CLI ....: yes
[DEBUG][Qt] 24.07.2017 11:11:00 [] (unknown:0) - Additional module paths ..: /Applications/Slicer.app/Contents/Extensions-26167/QuantitativeReporting/lib/Slicer-4.7/qt-scripted-modules, /Applications/Slicer.app/Contents/Extensions-26167/SlicerDevelopmentToolbox/lib/Slicer-4.7/qt-scripted-modules, /Applications/Slicer.app/Contents/Extensions-26167/DCMQI/lib/Slicer-4.7/cli-modules
[DEBUG][Python] 24.07.2017 11:11:10 [Python] (/Applications/Slicer.app/Contents/lib/Slicer-4.7/qt-scripted-modules/SubjectHierarchyPlugins/AbstractScriptedSubjectHierarchyPlugin.py:36) - Scripted subject hierarchy plugin registered: Annotations
[DEBUG][Python] 24.07.2017 11:11:10 [Python] (/Applications/Slicer.app/Contents/lib/Slicer-4.7/qt-scripted-modules/SubjectHierarchyPlugins/AbstractScriptedSubjectHierarchyPlugin.py:36) - Scripted subject hierarchy plugin registered: SegmentEditor
[DEBUG][Python] 24.07.2017 11:11:10 [Python] (/Applications/Slicer.app/Contents/lib/Slicer-4.7/qt-scripted-modules/SubjectHierarchyPlugins/AbstractScriptedSubjectHierarchyPlugin.py:36) - Scripted subject hierarchy plugin registered: SegmentStatistics
[DEBUG][Qt] 24.07.2017 11:11:02 [] (unknown:0) - Number of registered modules: 146
[CRITICAL][Stream] 24.07.2017 11:11:02 [] (unknown:0) - /Applications/Slicer.app/Contents/lib/Python/lib/python2.7/site-packages/numpy-1.13.1-py2.7-macosx-10.6-x86_64.egg/numpy/core/getlimits.py:82: RuntimeWarning: invalid value encountered in power
[CRITICAL][Stream] 24.07.2017 11:11:02 [] (unknown:0) -   self.resolution = float_to_float(float_conv(10) ** (-self.precision))
[DEBUG][Qt] 24.07.2017 11:11:03 [] (unknown:0) - Number of instantiated modules: 146
[WARNING][Qt] 24.07.2017 11:11:10 [] (unknown:0) - When loading module  "DICOMTID1500Plugin" , the dependency "DICOMRWVMPlugin" failed to be loaded.
[DEBUG][Qt] 24.07.2017 11:11:10 [] (unknown:0) - Number of loaded modules: 145
[DEBUG][Qt] 24.07.2017 11:11:10 [] (unknown:0) - Switch to module:  "Welcome"
[CRITICAL][Stream] 24.07.2017 11:11:10 [] (unknown:0) - Traceback (most recent call last):
[CRITICAL][Stream] 24.07.2017 11:11:10 [] (unknown:0) -   File "/Applications/Slicer.app/Contents/lib/Slicer-4.7/qt-scripted-modules/DICOM.py", line 85, in performPostModuleDiscoveryTasks
[CRITICAL][Stream] 24.07.2017 11:11:10 [] (unknown:0) -     DICOMLib.setDatabasePrecacheTags()
[CRITICAL][Stream] 24.07.2017 11:11:10 [] (unknown:0) -   File "/Applications/Slicer.app/Contents/lib/Slicer-4.7/qt-scripted-modules/DICOMLib/DICOMWidgets.py", line 36, in setDatabasePrecacheTags
[CRITICAL][Stream] 24.07.2017 11:11:10 [] (unknown:0) -     plugin = slicer.modules.dicomPlugins[pluginClass]()
[CRITICAL][Stream] 24.07.2017 11:11:10 [] (unknown:0) -   File "/Applications/Slicer.app/Contents/Extensions-26167/QuantitativeReporting/lib/Slicer-4.7/qt-scripted-modules/DICOMTID1500Plugin.py", line 23, in __init__
[CRITICAL][Stream] 24.07.2017 11:11:10 [] (unknown:0) -     self.rwvmPlugin = slicer.modules.dicomPlugins["DICOMRWVMPlugin"]()
[CRITICAL][Stream] 24.07.2017 11:11:10 [] (unknown:0) - KeyError: 'DICOMRWVMPlugin'
```

**Comment by pieper** [comment]
@fedorov no, I didn't do anything to the plugin loading, just the internals of the ScalarVolumePlugin.

is the DICOMRWVMPlugin actually there?

**Comment by fedorov** [comment]
looking into it...

**Comment by che85** [comment]
PETDICOMExtension is not building, that's why I added it manually from my checked out repo.

**Comment by che85** [comment]
related to https://github.com/QIICR/Slicer-PETDICOMExtension/pull/8

**Comment by fedorov** [comment]
I see. Extension manager actually gives a warning that a dependency is not available, but then proceeds to install QR, so I was confused. Sorry!

**Comment by pieper** [comment]
@che85 it looks like the SegmentationPlugin is returning a higher confidence than the ScalarVolumePlugin when trying to load the MR.

When you say "when trying to load the MR as a ScalarVolume" do you mean you uncheck the SegmentationPlugin and get the error from the ScalarVolume plugin?  That doesn't make sense because the error dialog looks like GDCM is trying to load a SEG (1 bit) and not an MR.

Sounds like a lot of other things are broken with QR and PET so if those get sorted out and there's a remaining problem with the ScalarVolumePlugin or the vtkITK layers let me know.

**Comment by che85** [comment]
I got the impression that the MR image before creation of the segmentation is not the same as after creating the segmentation. 

Just try the following:
1. remove the datasets from SlicerDICOMDatabase
2. open Quantitative Reporting and click button Retrieve and load test data
3. restart Slicer
4. load the test data from the SlicerDICOMDatabase

Loading works...

----
Do the following: 

5. Create segmentation
6. click Save Report
7. restart Slicer
8. try to load the MR image again. On my machine it will give the errors, but it shouldn't because it's an MR image and not a Segmentation

----

**Comment by fedorov** [comment]
@che85 the issue should be reproducible with the nightly to be able to debug it. Now that https://github.com/QIICR/Slicer-PETDICOMExtension/pull/8 is merged, let's test with tomorrow's nightly.

When I remove TID1500 plugin manually, I am able to use SEG plugin to load a segmentation without issues, and it does not return any loadables for the MR series I tested. It should not be possible for that plugin to return anything if Modality is not SEG in the input series, so I am not sure what is going on with your setup.

**Comment by che85** [comment]
@fedorov So after creating the segmentation and saving the report, how many series do you see in the SlicerDICOMDatabase? Here the seg doesn't even get listed.

**Comment by fedorov** [comment]
I could not test that with the nightly, because RWVM plugin is a requirement that is missing. I only tested the SEG plugin alone. I suggest we test TID1500 with the tomorrow nightly.

**Comment by che85** [comment]
If you want to test today, just add the DICOMPetExtension python plugins to your additional paths... otherwise yeah let's wait.

**Comment by fedorov** [comment]
I upgraded the OS, and haven't compiled Slicer recently. I wouldn't feel confident without full recompile, and the RWVM plugin depends on a CLI, so I prefer to wait and test in a clean setting.

**Comment by fedorov** [comment]
Having said that, CLI would not be invoked if we annotate a non-PET series, so I assume it should be ok to just add the path. You are probably right.

**Comment by lassoan** [comment]
Maybe the new DICOM object is exported with the same series instance UID as the MR series? That would explain why it does not show up in the browser (it is part of the MR series, which already shows up), and also why the series is misinterpreted as s segmentation object.

With the latest nightly version of Slicer you can easily copy to clipboard selected DICOM tags in all selected files, which may help with diagnosing the issue.

**Comment by fedorov** [comment]
I doubt that. We always create new instance UIDs for the newly created DICOM objects in dcmqi.

**Comment by pieper** [comment]
I agree with @lassoan that there is likely something funny with the UIDs.

**Comment by pieper** [comment]
Looks like the new dcmtk is messing up the ctkDICOMDatabase.

With a new build the interaction is very slow for me and the error log is filled with messages like this: 

SQL failed
 Bad SQL: INSERT OR REPLACE INTO TagCache VALUES(?,?,?)
Error text: disk I/O error Unable to fetch row

**Comment by pieper** [comment]
With a nightly build from a few days ago I get this:

<pre>
>>> i = '1.3.6.1.4.1.32722.99.99.267143140824284582760185651535851323805'
>>> s = "0010,0040"
>>> slicer.dicomDatabase.instanceValue(i,s)
u'M'
</pre>

But with my local build I get this (extra newline before the value)
<pre>
>>> s = "0010,0040"
>>> i = '1.3.6.1.4.1.32722.99.99.267143140824284582760185651535851323805'
>>> slicer.dicomDatabase.instanceValue(i,s)

u'M'
</pre>

It seems something changed in the DCMTK api that ctk needs to adapt to.

**Comment by pieper** [comment]
The extra newline thing is a regression apparently related to https://github.com/commontk/CTK/commit/3d30e4d1608d18e734925441702bc64084d85f8d.

No matter what I print out on the console I get an extra blank line:

<pre>
>>> s = 10
>>> s

10
</pre>

I'll report this on CTK issues and keep looking at the database thing.

**Comment by pieper** [comment]
Starting from a fresh database and loading a study I cannot replicate the slow behavior or the SQL errors I reported above.  The ctkDICOMDatabase layer seems to be working fine.  Let me know if anyone sees any errors.

**Comment by fedorov** [comment]
If we confirm that there is an issue with using an existing Slicer database, we should alert the users about the problem. 

I will test tomorrow with the nightly.

**Comment by lassoan** [comment]
The official explanation is "The SQLITE_IOERR result code says that the operation could not finish because the operating system reported an I/O error".

It seems that in practice "disk I/O error" [can be due to many things](http://sqlite.1065341.n5.nabble.com/SQLite-disk-I-O-error-td73098.html), such as [putting the system to sleep before the cache is flushed](https://github.com/clementine-player/Clementine/issues/4132), [almost full disk](https://sourceforge.net/p/mumble/discussion/492607/thread/46d94603/), etc.

If this happens frequently or reproducibly then we could have a look into changing some caching options (but it may impact performance, which is already quite slow) or alternative databases (would be nice but so much work...).

**Comment by pieper** [comment]
Yes, I think what I saw was a transient issue, but if it is reproducible and/or happens frequently fix it.

I'm still not sure what @che85 is seeing with the segmentation issue but it sounds like something related to the QR/dcmqi plugins so let us know what you find.

**Comment by michaelonken** [comment]
Guys, this is probably something in DCMTK: Jan told me about this commit many days ago: 

http://git.dcmtk.org/?p=dcmtk.git;a=commit;h=a9b866e50649d50df645fff61cdafb9121d1a7ef 

mentioning he fixed something with the parameter order (see iodcommn.cc) on the ::import() call. 

Looking into it (now, I was triggered by your description of the issue) it seems to me the commit breaks the API; the commit changes the parameter order of useSeries and useFoR. dcmqi "unfortunately" sets different values here, so the Frame of Reference (I guess) is now imported while the Series is not!

So I will undo that change. Could you easily try that? Just reverse useFoR and useSeries iodcommn.cc back to the old order, i.e. only in the .cc change the method signature to:

OFCondition DcmIODCommon::import(DcmItem& dataset,
                                 const OFBool readPatient,
                                 const OFBool readStudy,
                                 const OFBool readSeries, // was readFoR, needs to be readSeries
                                 const OFBool readFoR,   // was readSeries, needs to be readFoR
                                 const OFBool takeOverCharset)

**Comment by lassoan** [comment]
Does the issue present in current  nightly build?

If yes, then we should revert to the last working version of DCMTK immediately because latest stable is outdated and people rely on nightly builds.

**Comment by fedorov** [comment]
Andras, if I understand the description of Michael correctly, this issue would only affect dcmqi, which is an extension. I don't think the import function he mentioned is used anywhere else.

**Comment by fedorov** [comment]
Here is the line affected by the broken API: https://github.com/QIICR/dcmqi/blob/master/libsrc/ImageSEGConverter.cpp#L40

**Comment by lassoan** [comment]
OK, thank you!

**Comment by michaelonken** [comment]
@fedorov Let me provide a fix within this week; I will contact you via mail about the best option for the fix (the issue lays a bit deeper than obvious, fix is easy though)

**Comment by fedorov** [comment]
> Maybe the new DICOM object is exported with the same series instance UID as the MR series?

@lassoan you were right, and this is definitely was the case. It should be fixed in the PR above. Will test with the updated nightly tomorrow to see if there is anything else going on.

**Comment by che85** [comment]
I just checked with the recent nightly build. Everything works as expected. Closing this issue.

- [x] Creation works
- [x] Load works

![image](https://user-images.githubusercontent.com/10195822/28626241-7497697a-71ec-11e7-9e23-70fb35fb7d41.png)

**Comment by michaelonken** [comment]
For the record: The DCMTK import issue is now fixed in DCMTK upstream ([commit da17268](http://git.dcmtk.org/?p=dcmtk.git;a=commit;h=da172687b7cf59c4e377e4cdebf0d01ccca4f1ad)).

@fedorov I renamed the method to importHierarchy() in order to explicitly warn the user that the parameter order has changed. Otherwise he would maybe not notice the change but due to the renamed method he will now run into a compile error and read (hopefully) the related API documentation (which also warns explicitly about the issue).
